File: top-interview-questions-easy/Linked-List/palindrome-linked-list/234.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        fast = slow = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        prev = None
        while slow:
            next = slow.next
            slow.next = prev
            prev = slow
            slow = next

        while prev:
            if prev.val != head.val:
                return False
            prev = prev.next
            head = head.next
        return True

File: 234.回文链表/234.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */

// eslint-disable-next-line no-unused-vars
const isPalindrome = function(head) {
    if (!head || !head.next) return true

    // find middle
    let slow = head, fast = head, cacheHead = head
    while (fast && fast.next) {
        slow = slow.next
        fast = fast.next.next
    }

    // reverse link list
    let prev = null
    while (slow) {
        let next = slow.next
        slow.next = prev
        prev = slow
        slow = next
    }

    // judge palindrom
    let list1 = cacheHead, list2 = prev
    for (; list2; list1 = list1.next, list2 = list2.next) {
        if (list1.val !== list2.val) return false
    }
    return true
}
